File: pattern/Binary Tree Traversal/230. Kth Smallest Element in a BST(inorder)/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public static int count = 0;
    public static int res = 0;
    public int kthSmallest(TreeNode root, int k) {
        count = k;
        helper(root);
        return res;
    }

    public void helper ( TreeNode root){
        if(root.left != null){
            helper(root.left);
        }
        count--;
        if(count == 0){
            res = root.val;
            return;
        }
        if(root.right != null){
            helper(root.right);
        }
    }
}

File: pattern/Binary Tree Traversal/230. Kth Smallest Element in a BST(inorder)/Solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
  let result = 0;
  let count = k;
  function inorder(node) {
    if (node === null) {
      return;
    }
    inorder(node.left);
    count--;
    if (count == 0) {
      result = node.val;
      return;
    }
    inorder(node.right);
  }
  inorder(root);
  return result;
};
